server/rest/src/main/java/org/infinispan/rest/stream/CacheChunkedStream.java

```diff
@@ -0,0 +1,177 @@
+package org.infinispan.rest.stream;
+
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayDeque;
+import java.util.Queue;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.stream.ChunkedWriteHandler;
+import io.reactivex.rxjava3.subscribers.DefaultSubscriber;
+import org.infinispan.commons.marshall.WrappedByteArray;
+import org.infinispan.util.logging.Log;
+import org.infinispan.util.logging.LogFactory;
+import org.reactivestreams.Publisher;
+
+public abstract class CacheChunkedStream<T> implements ExtendedChunkedInput<ByteBuf>, CacheStreamProcessor<T> {
+   protected static final Log logger = LogFactory.getLog(CacheChunkedStream.class);
+
+   // Netty default value with `ChunkedStream`.
+   private static final int CHUNK_SIZE = 8192;
+
+   private final Publisher<T> publisher;
+   private final InternalStreamSubscriber subscriber;
+   private long offset;
+
+   private volatile boolean stuttered;
+   private volatile ChannelHandlerContext ctx;
+
+   public CacheChunkedStream(Publisher<T> publisher) {
+      this.publisher = publisher;
+      this.subscriber = new InternalStreamSubscriber();
+      this.publisher.subscribe(subscriber);
+   }
+
+   public static byte[] readContentAsBytes(Object content) {
+      if (content instanceof byte[]) return (byte[]) content;
+      if (content instanceof WrappedByteArray) {
+         return ((WrappedByteArray) content).getBytes();
+      }
+
+      return content.toString().getBytes(StandardCharsets.UTF_8);
+   }
+
+   @Override
+   public void close() { }
+
+   @Override
+   public ByteBuf readChunk(ChannelHandlerContext ctx) throws Exception {
+      return readChunk(ctx.alloc());
+   }
+
+   @Override
+   public ByteBuf readChunk(ByteBufAllocator allocator) throws Exception {
+      ByteBuf buf = readChunkOrStutter(allocator);
+      stuttered = buf == null;
+      return buf;
+   }
+
+   private ByteBuf readChunkOrStutter(ByteBufAllocator allocator) throws Exception {
+      boolean release = true;
+      ByteBuf buffer = allocator.buffer(CHUNK_SIZE);
+      try {
+         int i = 0;
+         while (i < CHUNK_SIZE) {
+            // Its possible the buffer was full before reading the element completely,
+            // in such case, we do not load another element, we continue from where we left.
+            if (!hasElement()) {
+               T item = subscriber.peek();
+               if (item == null) {
+                  // There is the possibility the stream is just empty and completed.
+                  // We need to process the `null` until is the end of the input.
+                  if (!subscriber.isCompleted() || isEndOfInput()) {
+                     break;
+                  }
+               }
+
+               setCurrent(item);
+            }
+
+            byte b;
+            while ((b = read()) != -1) {
```

Comment:
I believe this can overflow the buffer if the entry size is too large for the buffer.

TBH, I am not a huge fan of keeping the whole `read` code from before. I will try to make a commit to see if I can clean it up a bit.